gems/sorbet-runtime/lib/types/props/private/initializer_generator.rb

```diff
@@ -0,0 +1,114 @@
+# frozen_string_literal: true
+# typed: strict
+
+module T::Props
+  module Private
+
+    # Generates a specialized `initialize` implementation for a subclass of
+    # T::Props::Constructor.
+    #
+    # The basic idea is that we analyze the props and for each prop, generate
+    # the simplest possible logic as a block of Ruby source, so that...
+    module InitializerGenerator
+      extend T::Sig
+
+      sig do
+        params(
+          props: T::Hash[Symbol, T::Hash[Symbol, T.untyped]],
+          raise_on_missing_key: T::Boolean,
+        )
+        .returns(String)
+        .checked(:never)
+      end
+      def self.generate(props, raise_on_missing_key:)
+        parts = props.map do |prop, rules|
+          # All of these strings should already be validated (directly or
+          # indirectly) in `validate_prop_name`, so we don't bother with a nice
+          # error message, but we double check here to prevent a refactoring
+          # from introducing a security vulnerability.
+          raise unless T::Props::Decorator::SAFE_NAME.match?(prop.to_s)
+
+          ivar_name = rules.fetch(:accessor_key).to_s
+          raise unless ivar_name.start_with?('@') && T::Props::Decorator::SAFE_NAME.match?(ivar_name[1..-1])
+
+          hash_key = prop.inspect
+          need_type_check = T::Props::Utils.need_type_check?(rules)
+
+          present_key_value = <<~RUBY.strip
+            hash[#{hash_key}]
+          RUBY
+          present_key_value = wrap_with_typecheck(present_key_value, hash_key) if need_type_check
+
+          if rules.key?(:factory)
+            missing_key_value = <<~RUBY.strip
+              self.class.class_exec(&decorator.props[#{hash_key}].fetch(:factory))
+            RUBY
+            missing_key_value = wrap_with_typecheck(missing_key_value, hash_key) if need_type_check
+          elsif rules.key?(:default)
+            missing_key_value = generate_simple_default(rules.fetch(:default)) || <<~RUBY.strip
+              T::Props::Utils.deep_clone_object(decorator.props[#{hash_key}].fetch(:default))
+            RUBY
+            # FIXME: Ideally we'd check here that the default is actually a valid
+            # value for this field, but existing code relies on the fact that we don't.
+            #
+            # :(
+            #
+            # missing_key_value = wrap_with_typecheck(missing_key_value, hash_key) if need_type_check
+          elsif T::Props::Utils.required_prop?(rules) && raise_on_missing_key
+            missing_key_value = <<~RUBY.strip
+              raise ArgumentError.new("Missing required prop `#{prop}` for class `\#{self.class.name}`")
+            RUBY
+          else
+            missing_key_value = <<~RUBY.strip
+              nil
+            RUBY
+          end
+
+          <<~RUBY.strip
+            #{ivar_name} = if hash.key?(#{hash_key})
```

Comment:
I dislike that we're doing two hash lookups for each key in the normal case, but I'm not sure there's an alternative that's simple enough to be worthwhile.

You could do something like:
```
value = present_key_value
if value.nil? && !hash.key?(#{hash_key})
  value = missing_key_value
else
  result += 1
end
```

... but I think that might change what error is raised for missing required props in a way that we don't want. Maybe worth benchmarking and then seeing if there's a reasonable solution for that problem if it's meaningfully faster?

Also, this makes me realize I'm not actually sure what behavior we expect or want in the case where a required prop with `checked: :never` is missing from the constructor.

Reply:
Great call! I pushed up a commit which does basically what you suggested and optimizes for the case where the value is present. See the updated PR description with the performance benchmarks, it mostly improves performance across the board 🚀 although it flip/flops the `checked(:never)` runtimes but in an expected and desirable way - when props are mostly given and non-null it's faster.